### PR TITLE
[PIN-1290] Adding route POST /eservices for catalog-process

### PR DESCRIPTION
**Pull request by bloodsky**

### CODE REVIEW

File: src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/CatalogApiServiceImpl.scala [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+  override def createEService(eServiceSeed: EServiceSeed)(implicit
+    contexts: Seq[(String, String)],
+    toEntityMarshallerOldEService: ToEntityMarshaller[OldEService],
+    toEntityMarshallerProblem: ToEntityMarshaller[Problem]
+  ): Route = {
+    val result: Future[OldEService] = for {
+      response <- catalogProcess.createEService(eServiceSeed.toProcess)(contexts)
+      converted = response.toApi
+    } yield converted
```
Comment (TonioGela) [inline src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/CatalogApiServiceImpl.scala]: I would instead prefer something like this, as the for comprehension is not strictly necessary in this case:

```suggestion
    def result: Future[OldEService] = catalogProcess.createEService(eServiceSeed.toProcess)(contexts).map(_.toApi)
```

or at least to move the mapping in the yield like: 

```suggestion
    val result: Future[OldEService] = for {
      response <- catalogProcess.createEService(eServiceSeed.toProcess)(contexts)
    } yield response.toApi
```

as it won't be sent two times to two different execution contexts

File: src/main/scala/it/pagopa/interop/backendforfrontend/service/CatalogProcessService.scala [hunk 2]
```diff
@@ -7,6 +7,7 @@ import scala.concurrent.Future
 
 trait CatalogProcessService {
 
+  def createEService(eServiceSeed: EServiceSeed)(implicit contexts: Seq[(String, String)]): Future[OldEService]
   def getEServices(
```
Comment (TonioGela) [inline src/main/scala/it/pagopa/interop/backendforfrontend/service/CatalogProcessService.scala:12]: can you add a \n?

File: src/main/scala/it/pagopa/interop/backendforfrontend/service/types/CatalogProcessServiceTypes.scala [hunk 3]
```diff
@@ -7,13 +7,108 @@ import it.pagopa.interop.backendforfrontend.model.EServiceDescriptorState._
...
+      name = coav.name,
+      description = coav.description,
+      explicitAttributeVerification = coav.explicitAttributeVerification
+    )
+  }
+  implicit class CatalogOldAttributeConverter(private val coa: CatalogProcessOldAttribute)            extends AnyVal {
+    def toApi: OldAttribute =
+      OldAttribute(single = coa.single.map(_.toApi), group = coa.group.nested.map(_.toApi).value)
+  }
+  implicit class CatalogOldAttributesConverter(private val coas: CatalogProcessOldAttributes)         extends AnyVal {
```
Comment (TonioGela) [inline src/main/scala/it/pagopa/interop/backendforfrontend/service/types/CatalogProcessServiceTypes.scala]: There's a lack of \n between classes here, I won't mark it all

**Review by TonioGela — CHANGES_REQUESTED** [review]

File: src/main/resources/interface-specification.yml [hunk 4]
```diff
@@ -1574,6 +1574,37 @@ paths:
...
+  /eservices:
+    parameters:
+      - $ref: '#/components/parameters/CorrelationIdHeader'
+      - $ref: '#/components/parameters/IpAddress'
+    post:
+      security:
+        - bearerAuth: [ ]
+      tags:
+        - catalog
+      summary: Create a new e-service
```
Comment (galales) [inline src/main/resources/interface-specification.yml]: Let's use the same naming everywhere
```suggestion
      summary: Create a new EService
```

File: src/main/resources/interface-specification.yml [hunk 5]
```diff
@@ -1574,6 +1574,37 @@ paths:
...
+      operationId: createEService
+      requestBody:
+        description: An E-Service seed
```
Comment (galales) [inline src/main/resources/interface-specification.yml]: same
```suggestion
        description: An EService seed
```

File: src/main/resources/interface-specification.yml [hunk 6]
```diff
@@ -1574,6 +1574,37 @@ paths:
...
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/OldEService'
+        '400':
+          description: Invalid input
+          content:
+            application/problem+json:
+              schema:
+                $ref: '#/components/schemas/Problem'
```
Comment (galales) [inline src/main/resources/interface-specification.yml]: rate limiting headers needed

File: src/main/resources/interface-specification.yml [hunk 7]
```diff
@@ -3431,6 +3462,184 @@ components:
...
+          type: string
+          format: uuid
+        name:
+          type: string
+          minLength: 5
+          maxLength: 60
+        description:
+          type: string
+          minLength: 10
+          maxLength: 250
```
Comment (galales) [inline src/main/resources/interface-specification.yml]: constraints should not be in the BFF, we agreed to centralize them only on process services
```suggestion
        description:
          type: string
```

File: src/main/resources/interface-specification.yml [hunk 8]
```diff
@@ -1574,6 +1574,37 @@ paths:
...
+            schema:
+              $ref: '#/components/schemas/EServiceSeed'
+        required: true
+      responses:
+        '200':
+          description: EService created
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/OldEService'
```
Comment (galales) [inline src/main/resources/interface-specification.yml]: (confirmed by @Carminepo2 ) we need to return just the id of the new eservice, you can use the `CreatedResource` model

Btw, models named `Old*` are just intended as a temporary measure in the catalog process.
Names in the BFF should not be named with `Old` prefix (and these specific models are not needed here)

File: src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/CatalogApiServiceImpl.scala [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+import it.pagopa.interop.backendforfrontend.server.impl.Main.loggerTI.canLogEv
+import it.pagopa.interop.backendforfrontend.service.CatalogProcessService
+import it.pagopa.interop.backendforfrontend.service.types.CatalogProcessServiceTypes.CatalogEServiceSeedConverter
+import it.pagopa.interop.backendforfrontend.service.types.CatalogProcessServiceTypes.CatalogOldEServiceConverter
+import it.pagopa.interop.commons.logging.ContextFieldsToLog
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Success
+
+final case class CatalogApiServiceImpl(catalogProcess: CatalogProcessService)(implicit ec: ExecutionContext)
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/CatalogApiServiceImpl.scala]: in the BFF already exists the `EServicesApiServiceImpl`, you should move this logic there

File: src/main/resources/interface-specification.yml [hunk 10]
```diff
@@ -1574,6 +1574,37 @@ paths:
...
               description: Time interval in milliseconds. Allowed requests will be constantly replenished during the interval. At the end of the interval the max allowed requests will be available
+  /eservices:
+    parameters:
+      - $ref: '#/components/parameters/CorrelationIdHeader'
+      - $ref: '#/components/parameters/IpAddress'
+    post:
+      security:
+        - bearerAuth: [ ]
+      tags:
+        - catalog
```
Comment (galales) [inline src/main/resources/interface-specification.yml]:
```suggestion
        - eservices
```

File: src/main/resources/interface-specification.yml [hunk 11]
```diff
@@ -3431,6 +3462,184 @@ components:
...
+        description:
+          type: string
+        technology:
+          $ref: '#/components/schemas/EServiceTechnology'
+        attributes:
+          $ref: '#/components/schemas/OldAttributes'
+        descriptors:
+          type: array
+          items:
+            $ref: '#/components/schemas/EServiceDescriptor'
```
Comment (galales) [inline src/main/resources/interface-specification.yml]: you probably do not need these models

**Review by galales — CHANGES_REQUESTED** [review]

**Review by galales — APPROVED** [review]
🚀 
Just remember that [this](https://github.com/pagopa/interop-be-backend-for-frontend/pull/82/files#diff-b9cc9357f75e7dd8cdee77916533ecee2bab1e58a2aa0c2b823e361346eeaa22R23) will slightly change once the corresponding endpoint in the catalog process will be updated

**Review by TonioGela — APPROVED** [review]

**Review by TonioGela — APPROVED** [review]

**Review by galales — APPROVED** [review]

**Pull request merged**